File: Output/javascript/medium/AllNodesDistanceKInBinaryTree.js
/**https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/ */
//Given the root of a binary tree, the value of a target node target, and an integer k, return an array of the values of all nodes that have a distance k from the target node.
//You can return the answer in any order.


/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} target
 * @param {number} k
 * @return {number[]}
 */
var distanceK = function(root, target, k) {
    let map = {};
    let result = [];
    let queue = [];
    let visited = new Set();
    queue.push(root);
    visited.add(root);
    while(queue.length > 0){
        let node = queue.shift();
        if(node.left != null){
            map[node.left.val] = node;
            queue.push(node.left);
            visited.add(node.left);
        }
        if(node.right != null){
            map[node.right.val] = node;
            queue.push(node.right);
            visited.add(node.right);
        }
    }
    queue.push(target);
    visited.add(target);
    let level = 0;
    while(queue.length > 0){
        let size = queue.length;
        for(let i = 0; i < size; i++){
            let node = queue.shift();
            if(level == k){
                result.push(node.val);
            }
            if(node.left != null && !visited.has(node.left)){
                queue.push(node.left);
                visited.add(node.left);
            }
            if(node.right != null && !visited.has(node.right)){
                queue.push(node.right);
                visited.add(node.right);
            }
            if(map[node.val] != null && !visited.has(map[node.val])){
                queue.push(map[node.val]);
                visited.add(map[node.val]);
            }
        }
        level++;
    }
    return result;
    
};

    

File: Tool/PreUtil/src/output/python/medium/AllNodesDistanceKInBinaryTree.py
#https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
#Given the root of a binary tree, the value of a target node target, and an integer k, return an array of the values of all nodes that have a distance k from the target node.
#You can return the answer in any order.


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def distanceK(self, root, target, k):
        """
        :type root: TreeNode
        :type target: TreeNode
        :type k: int
        :rtype: List[int]
        """       
        def dfs(node, parent):
            if node:
                node.parent = parent
                dfs(node.left, node)
                dfs(node.right, node)
        dfs(root, None)
        queue = collections.deque([(target, 0)])
        seen = {target}
        while queue:
            if queue[0][1] == k:
                return [node.val for node, d in queue]
            node, d = queue.popleft()
            for nei in (node.left, node.right, node.parent):
                if nei and nei not in seen:
                    seen.add(nei)
                    queue.append((nei, d+1))
        return []

        

    